torch/csrc/distributed/c10d/reducer.cpp

```diff
@@ -631,7 +631,11 @@ void Reducer::autograd_hook(size_t index) {
     });
   }
 
-  if (static_graph_first_iteration()) {
+  if (static_graph_first_iteration() && !static_graph_first_iteration_cb_enqueued_) {
```

Comment:
will this still work if the first parameter (or all parameters) are not used in one rank? Or do we never expect delay_all_reduce() to be combined with unused parameters?

If unused parameters is a concern here, will it be more generic if we do `_find_tensors` on inputs or outputs and call queue callback there?

Reply:
This works if the first param is not used on a particular rank, and the unittest that has a model with a single param confirms this.

We'd have to investigate a bit more in the case all params are not used on a particular rank. We can try the following case:

def forward(x):
    if rank == 0:
        x = self.lin(x)
    return x

and check whether the current and new methods of delaying allreduce work or not.